### PR TITLE
re #2104: ivarator limit

### PR DESCRIPTION
**Pull request by ivakegg**
* Added max ivarator terms configuration property
* Enforce max ivarator term limit in validateQuerySize
* Added test

### CODE REVIEW

**Review by keith-ratcliffe — DISMISSED** [review]
LGTM

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 1]
```diff
@@ -672,6 +678,17 @@ public static void validateQuerySize(String lastOperation, final JexlNode queryT
             throw new DatawaveFatalQueryException(qe);
         }
 
+        // now check whether we are over the ivarator limit
+        if (maxIvaratorThreshold >= 0) {
+            NodeTypeCount nodeCount = NodeTypeCountVisitor.countNodes(queryTree);
+            int totalIvarators = nodeCount.getTotal(ExceededValueThresholdMarkerJexlNode.class) + nodeCount.getTotal(ExceededOrThresholdMarkerJexlNode.class);
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Seems like the NodeTypeCountVisitor should optionally accept a list of class names as a filter..

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 1 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: The visitor still needs to visit every node.  To add an additional check at every node to determine whether it is something we are interested in actually may be more expensive, but that is merely a hypothesis.  I will try to test that out.

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Well, after much testing it appears that adding a set to test against is neglegible in terms of addition time.  Adding that in.

**Review by apmoriarty — APPROVED** [review]

**Review by keith-ratcliffe — APPROVED** [review]

**Review by MiguelRicardos — APPROVED** [review]

**Pull request merged**